corehq/apps/users/management/commands/migrate_reprs.py

```diff
@@ -0,0 +1,40 @@
+import logging
+
+from django.core.management.base import BaseCommand
+from django.db.models import Q
+from corehq.apps.users.models import UserHistory
+from corehq.apps.users.util import cached_user_id_to_username
+
+from corehq.apps.users.util import SYSTEM_USER_ID
+from corehq.util.log import with_progress_bar
+
+
+class Command(BaseCommand):
+    help = "Adds user_repr and changed_by_repr to UserHistory log if not already present"
+
+    def handle(self):
+        records = UserHistory.objects.filter(Q(user_repr__isnull=True) | Q(changed_by_repr__isnull=True))
+        for user_history in with_progress_bar(records.order_by('pk').iterator(), records.count()):
+            try:
+                migrate(user_history)
+            except Exception as e:
+                logging.error(f"{user_history.pk}: {e}")
+
+
+def migrate(user_history):
+    was_change = False
+    if user_history.changed_by == SYSTEM_USER_ID:
+        changed_by_repr = SYSTEM_USER_ID
+    else:
+        changed_by_repr = cached_user_id_to_username(user_history.changed_by)
+
+    user_repr = cached_user_id_to_username(user_history.user_id)
+
+    if user_history.changed_by_repr != changed_by_repr or user_history.user_repr != user_repr:
+        was_change = True
+    if changed_by_repr:
+        user_history.changed_by_repr = changed_by_repr
+    if user_repr:
+        user_history.user_repr = user_repr
+    if was_change:
+        user_history.save()
```

Comment:
I don't think it matters a ton in this case, but there's a slight weirdness to the way this is written.  I think what you want to do conceptually is:

> If the model has changed, save it

But this only actually changes the model if one of the new values is non-null.  You could imagine the existing model has `user_repr` set, but this migration script sees `user_repr` as `None` (maybe the user has since been deleted, say).  In that case, `was_change` would be set to True, because there's a difference, but `user_repr` wouldn't actually be set to `None` (which is probably correct behavior), so the save is unnecessary.

An alternative is to push the two together:

```suggestion
    if changed_by_repr and user_history.changed_by_repr != changed_by_repr:
        user_history.changed_by_repr = changed_by_repr
        was_change = True
    if user_repr and user_history.user_repr != user_repr:
        user_history.user_repr = user_repr
        was_change = True
    if was_change:
        user_history.save()
```

Like I said, it doesn't matter much in this case, as that extra save would probably happen rarely if ever, and it is unlikely to be a performance concern.  It's more that this type of thing can be difficult to refactor down the road, since future devs might not know whether something relies on that save happening in this edge case, so it's nice to have really tight, explicit logic.

Reply:
Ah, i see. Doing the checks this way gets you closer to the main question of whether or not a change has occurred.   Thanks, making that change.